### ISSUE TITLE
DeepChem Development Roadmap

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#343]
I thought it would be useful to lay out a rough development roadmap for DeepChem to give users a sense of the relatively stability of parts of the DeepChem API. Since DeepChem is alpha software, users should be willing to put up with some amount of breakage, but I'd like to make things as painless as possible by explicitly noting which parts of API are fair game for future refactoring.

- `dc.data` - **Relatively Stable**: The abstract `Dataset` API seems to be holding up well, so I don't anticipate any major changes here. It is likely that we'll continue tweaking things here and there though (I think `NumpyDataset` is missing a little functionality and `DiskDataset` may be simplified a bit. Might also add other subclasses as need arises).
- `dc.dock` - **Highly Unstable**: This is the newest part of DeepChem and the API is not at all settled. Any and all parts of API may change without warning for foreseeable future.
- `dc.feat` -  **Relatively Stable**: While I anticipate us adding on new `Featurizer` objects, I don't think we will make any major changes to existing code. The one exception is `GridFeaturizer`, which will likely be refactored at some point.
- `dc.hyper` - **Deprecated**: The current hyperparameter search framework is a little kludgey. I anticipate that we may clean this up, or remove outright (it's easy enough to do hyperparam search, so we might just have a basic example instead). This isn't high on my priorities list, so might stick around for a few months or more though.
- `dc.metrics` - **Stable**: No major changes anticipated. May add a couple new metrics as need arises.
- `dc.models` - **Unstable**: The DeepChem model building framework is outdated and a little kludgey. I'd like to support `Keras` style model construction, but I'd like to be able to do it entirely within DeepChem without Keras (#296). Refactoring `dc.models` will be tricky though, so won't start until we solidify the benchmark suite (so we can refactor without harming performance).
- ` dc.nn` - **Relatively Stable**: I don't anticipate removing primitives, but we will likely add a good number more primitives moving forward.
- `dc.splits` - **Relatively Stable**: No major refactoring anticipated. However, @joegomes has reported that the `RandomSplitter` may have issues (doesn't behave deterministically even when seed is set), so likely will need a couple bugfixes.
- `dc.trans` - **Relatively Stable**: No major refactoring anticipated, but will likely add more transformers and debug existing transformers.
- `dc.utils` - **Deprecated**:  This submodule is currently a dumping bin. I'd recommend avoiding using anything in here. I'll try to clean-up here. Will likely greatly trim or outright remove this module.

**Comment by proteneer** [comment]
On the splitting side, getting the RandomStratifiedSplitters to be
consistent with the other classes would be another improvement. In
addition, making computing statistics more consistent would be nice for
graph mol features.

**Comment by rbharath** [comment]
@proteneer :+1:  for `RandomStratifiedSplitter`. PR welcome for that one.

Are you referring to `compute_feature_statistics`? I believe that all the feature statistics code was removed in #317, so hopefully shouldn't be an issue any more.

**Comment by proteneer** [comment]
Ah,

I need to update to master then wrt the feature statistic arg

**Comment by rbharath** [comment]
As an update here, I've started refactoring `dc.models`. The goal is to support expressive, Keras-style model building directly in DeepChem, but without losing the powerful flexibility of raw Tensorflow. As a first step, I've merged in #357 which removes the Keras dependency from DeepChem by copying in all Keras code which DeepChem relies upon into `deepchem.nn`. I could then simplify the code by removing support for Keras features we don't need (configs, masks, sparse tensors).

My target is to solve #292 and #349 over the next couple weeks by more tightly integrating the new code copied over from Keras with the existing Tensorflow framework. This should lay the ground work for #359, which will enable Keras-style model building with DeepChem.
